contrib/babelfishpg_tsql/src/pl_exec.c

```diff
@@ -4597,10 +4597,79 @@ exec_stmt_execsql(PLtsql_execstate *estate,
 	bool		enable_txn_in_triggers = !pltsql_disable_txn_in_triggers;
     StringInfoData query;
 	Oid			current_user_id = GetUserId();
+	PLExecStateCallStack *top_es_entry;
+	bool		need_path_reset = false;
+	char		*cur_dbname = get_cur_db_name();
+	/* fetch current search_path */
+	List 		*path_oids = fetch_search_path(false);
+	char 		*old_search_path = flatten_search_path(path_oids);
+	char 		*new_search_path;
+	char 		*physical_schema;
+	char		*dbo_schema;
 
 	if (stmt->is_cross_db)
 		SetCurrentRoleId(GetSessionUserId(), false);
 
+ 	top_es_entry = exec_state_call_stack->next;
+	if (stmt->is_dml)
+	{
+		while(top_es_entry != NULL)
+		{
+			/* traverse through the estate stack. If the occurrence of
+			 * exec in the call stack, update the search path.
+			 */
+			if(top_es_entry->estate && top_es_entry->estate->err_stmt &&
+				top_es_entry->estate->err_stmt->cmd_type == PLTSQL_STMT_EXEC &&
+					top_es_entry->estate->schema_name != NULL)
+ 				{
+ 					if (top_es_entry->estate->db_name == NULL)
+ 					{
+ 						physical_schema = get_physical_schema_name(cur_dbname, top_es_entry->estate->schema_name);
+ 						dbo_schema = get_dbo_schema_name(cur_dbname);
+ 					}
+ 					else
+ 					{
+ 						physical_schema = get_physical_schema_name(top_es_entry->estate->db_name, top_es_entry->estate->schema_name);
+ 						dbo_schema = get_dbo_schema_name(top_es_entry->estate->db_name);
+ 					}
+ 					new_search_path = psprintf("%s, %s, %s", physical_schema, dbo_schema, old_search_path);
+ 					/* Add the schema where the object is referenced and dbo schema to the new search path */
+ 					(void) set_config_option("search_path", new_search_path,
+ 									PGC_USERSET, PGC_S_SESSION,
+ 									GUC_ACTION_SAVE, true, 0, false);
+ 					need_path_reset = true;
+ 					break;
+ 				}
+ 			else
+ 				top_es_entry = top_es_entry->next;
+ 		}
+ 	}
+	 
+	if (stmt->is_ddl)
+	{
+		while(top_es_entry != NULL)
+		{
+			/* traverse through the estate stack. If the occurrence of
+			 * exec in the call stack, update the search path.
+			 */
+			if(top_es_entry->estate && top_es_entry->estate->err_stmt &&
+				top_es_entry->estate->err_stmt->cmd_type == PLTSQL_STMT_EXEC &&
+					top_es_entry->estate->db_name != NULL)
+				{
+					dbo_schema = get_dbo_schema_name(top_es_entry->estate->db_name);
+					new_search_path = psprintf("%s, %s", dbo_schema, old_search_path);
```

Comment:
why DBO for DDLs, PR description seems to imply otherwise?

Reply:
For DDL (CREATE TABLE), an object created in a proc will be in the default schema for the caller of the proc. Modified it in the next commit.